pkg/routing/signal.go

```diff
@@ -94,49 +93,248 @@ func (r *signalClient) StartParticipantSignal(
+		Writer:         signalRequestMessageWriter{},
+		CloseOnFailure: true,
+	})
 	resChan := NewDefaultMessageChannel()
 
 	go func() {
 		r.active.Inc()
 		defer r.active.Dec()
 
-		var err error
-		for msg := range stream.Channel() {
-			if err = resChan.WriteMessage(msg.Response); err != nil {
-				break
+		err = CopySignalStreamToMessageChannel[*rpc.RelaySignalRequest, *rpc.RelaySignalResponse](
+			stream,
+			resChan,
+			signalResponseMessageReader{},
+			r.config,
+		)
+		l.Debugw("participant signal stream closed", "error", err)
+
+		resChan.Close()
+	}()
+
+	return connectionID, sink, resChan, nil
+}
+
+type signalRequestMessageWriter struct{}
+
+func (e signalRequestMessageWriter) WriteOne(seq uint64, msg proto.Message) *rpc.RelaySignalRequest {
+	return &rpc.RelaySignalRequest{
+		Seq:     seq,
+		Request: msg.(*livekit.SignalRequest),
+	}
+}
+
+func (e signalRequestMessageWriter) WriteMany(seq uint64, msgs []proto.Message) *rpc.RelaySignalRequest {
+	r := &rpc.RelaySignalRequest{
+		Seq:      seq,
+		Requests: make([]*livekit.SignalRequest, 0, len(msgs)),
+	}
+	for _, m := range msgs {
+		r.Requests = append(r.Requests, m.(*livekit.SignalRequest))
+	}
+	return r
+}
+
+type signalResponseMessageReader struct{}
+
+func (e signalResponseMessageReader) Read(rm *rpc.RelaySignalResponse) ([]proto.Message, error) {
+	msgs := make([]proto.Message, 0, len(rm.Responses)+1)
+	if rm.Response != nil {
+		msgs = append(msgs, rm.Response)
+	}
+	for _, m := range rm.Responses {
+		msgs = append(msgs, m)
+	}
+	return msgs, nil
+}
+
+type RelaySignalMessage interface {
+	proto.Message
+	GetSeq() uint64
+}
+
+type SignalMessageWriter[SendType RelaySignalMessage] interface {
+	WriteOne(seq uint64, msg proto.Message) SendType
+	WriteMany(seq uint64, msgs []proto.Message) SendType
+}
+
+type SignalMessageReader[RecvType RelaySignalMessage] interface {
+	Read(msg RecvType) ([]proto.Message, error)
+}
+
+func CopySignalStreamToMessageChannel[SendType, RecvType RelaySignalMessage](
+	stream psrpc.Stream[SendType, RecvType],
+	ch *MessageChannel,
+	reader SignalMessageReader[RecvType],
+	config config.SignalRelayConfig,
+) error {
+	r := &signalMessageReader[SendType, RecvType]{
+		reader: reader,
+		config: config,
+	}
+	for msg := range stream.Channel() {
+		var res []proto.Message
+		res, err := r.Read(msg)
+		if err != nil {
+			return err
+		}
+		for _, r := range res {
+			if err = ch.WriteMessage(r); err != nil {
+				return err
 			}
-			for _, res := range msg.Responses {
-				if err = resChan.WriteMessage(res); err != nil {
-					break
-				}
+		}
+	}
+	return stream.Err()
+}
+
+type signalMessageReader[SendType, RecvType RelaySignalMessage] struct {
+	seq    uint64
+	reader SignalMessageReader[RecvType]
+	config config.SignalRelayConfig
+}
+
+func (r *signalMessageReader[SendType, RecvType]) Read(msg RecvType) ([]proto.Message, error) {
+	res, err := r.reader.Read(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.config.MinVersion >= 1 {
```

Comment:
See the version config usage.  Guess in this case it will not be possible to derive from protocol. Should we think about versioning in protocol messages itself? Definitely a larger discussion, but wanted to check if we should take this opportunity to standardise this.

For client <-> server, we use a single number.  Maybe, PsRPC already has something for signal relay client <-> server?

Reply:
the client could send the supported version in the open message and the server could send back its version before switching to reliable streaming. i considered this but the tradeoffs didn't make sense to me.